packages/stream/index.js

```diff
@@ -115,6 +115,91 @@ export const makePipe = () => {
 };
 harden(makePipe);
 
+/**
+ * @template TRead
+ * @template TWrite
+ * @template TReadReturn
+ * @template TWriteReturn
+ * @param {import('./types.js').Stream<TWrite, TRead, TWriteReturn, TReadReturn>} writer
+ * @param {import('./types.js').Stream<TRead, TWrite, TReadReturn, TWriteReturn>} reader
+ * @param {TWrite} primer
+ */
+export const pump = async (writer, reader, primer) => {
+  /** @param {Promise<IteratorResult<TRead, TReadReturn>>} promise */
+  const tick = promise =>
+    promise.then(
```

Comment:
Since we are writing this for a Hardened JS environment (thanks!), the abstractions should be more defensive. This synchronous call to `.then` enables the provider of the alleged promise to mount a reentrancy attack. Can we use `E.when`? Depends on how you want to layer endo, but I imagine eventual-send is lower level than stream.

Likewise for other calls to `.then`

Even an `await`, for all its flaws, is safer than a direct call to an unverified `.then`.

Btw, `E.when` accumulates deep stacks for debugging. Nether `.then` nor `await` do.

Reply:
I’ve hesitated to introduce that dependency, but I see now that E doesn’t actually entrain CapTP or Marshal, just the HandledPromise/Promise.delegate shim.

Streams are going to have to entrain a dependency on eventual-send when we want streams to have hooks for the transport layer. So I’m okay with bringing that dependency in.

Reply:
Changing `promise.then` to `E.when(promise` suggests we’re expecting an adversarial reader or writer. Do we need to arrange a similar defense for the stream methods, `next`, `throw`, `return`?

Reply:
If cheap enough, yes. Most abstractions should be written defensively by default. It is much easier to understand how to use them as a component within adversarial code. Every abstraction that isn't defensive should be marked somehow as fragile, so that it is never released across a trust boundary. Ideally, we would even check and prevent such leakage automatically.

What would it take for these to be defensive?

Reply:
Probably exactly the same mechanism as `E.when`. We could wrap any stream such that all its methods get called in turns that resolve the previously returned promise.

Reply:
That would be expensive enough that I would like to defer the choice to the user. https://github.com/endojs/endo/issues/1003

Reply:
Agreed to defer discussion. It should be legal to call `.next` without waiting for a previous promise to resolve.